### PR TITLE
Add State Generator

### PR DESCRIPTION
**Pull request by nisdas**
This PR will implement a state generator, which will allow us to revert back to a previous state in case of forks or re-orgs.

- [x] Add methods in db to save last finalized state and retrieve them.
- [x] Add a state generator which generates state from the last finalized epoch till
the desired slot/epoch.
- [x] Add relevant tests.

### CODE REVIEW

**Comment by terenc3t** [comment]
Is this required for test net?

File: beacon-chain/blockchain/stateGen/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,26 @@
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
+
+go_library(
+    name = "go_default_library",
+    srcs = ["stateGen.go"],
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/stateGen/BUILD.bazel]: all lowercase filenames, source: https://rakyll.org/style-packages/

File: beacon-chain/blockchain/stateGen/stateGen.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+// GenerateStateFromSlot generates state from the last finalized epoch till the specified slot.
+func GenerateStateFromSlot(ctx context.Context, db *db.BeaconDB, slot uint64) (*pb.BeaconState, error) {
+	fState, err := db.FinalizedState()
+	if err != nil {
+		return nil, err
+	}
+
+	if fState.Slot >= slot {
+		return nil, fmt.Errorf("requested slot is lower than or equal to the current slot in the finalized beacon state."+
+			" Current finalized slot in state %d but was requested %d",
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/stateGen/stateGen.go]: A new line like this shouldn't start with a space. It should be like this:

```go
  str := "some really long " +
     "string that goes over " +
     "multiple lines"
```

File: beacon-chain/blockchain/stateGen/stateGen.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+		if !exists {
+			fState, err = state.ExecuteStateTransition(
+				ctx,
+				fState,
+				nil,
+				root,
+				true, /* sig verify */
+			)
+			if err != nil {
+				return nil, fmt.Errorf("could not execute state transition %v", err)
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/blockchain/stateGen/stateGen.go]: Error shouldn't start with capital letter

File: beacon-chain/db/block.go [hunk 4]
```diff
@@ -155,12 +155,12 @@ func (db *BeaconDB) BlockBySlot(slot uint64) (*pb.BeaconBlock, error) {
 
 		blockRoot := mainChain.Get(slotEnc)
 		if blockRoot == nil {
-			return nil
+			return fmt.Errorf("block not found for slot %d", slot)
```
Comment (prestonvanloon) [inline beacon-chain/db/block.go]: Should this be an error? I think returning nil is acceptable if some object is not present in a datastore

Comment (prestonvanloon) [reply beacon-chain/db/block.go]: If we want to return an error, then we should return a package level constant error not found

```
const ErrNotFound = errors.New("not found")

...

if blockRoot == nil {
    return ErrNotFound
}
```

This way you can do special handling like
```
err := db.BlockBySlot(slot)
if err == db.ErrNotFound {
  // handle not found err
  return
}
if err != nil {
  // handle some other error
  log.Error(err)
}
...
```

File: beacon-chain/blockchain/stategenerator/state_generator_test.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
+package stategenerator
+
+import (
+	"context"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/prysmaticlabs/prysm/shared/params"
+
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/stategenerator/state_generator_test.go]: Remove space

File: beacon-chain/blockchain/stategenerator/state_generator_test.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
+package stategenerator
+
+import (
+	"context"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/stategenerator/state_generator_test.go]: Remove space

**Pull request merged**